template.tpl

```diff
@@ -0,0 +1,1704 @@
+          if (ruleIdx >= 0) {
+            const setCtxKey = entityRules[ruleIdx].mappedKey;
+            setFromPath(setCtxKey, ctxVal, tagConfig, returnObj);
+          } else if (tagConfig.includeUnmappedEntities) {
+            addProperty(prop, ctxVal, tagConfig, NEST.unmappedCtx, returnObj);
+          }
+        }
+      }
+    }
+  }
+  return returnObj;
+};
+
+/*
+ * Adds client event properties to the payload according to tag configuration.
+ *
+ * @param evData {Object} - the client event data
+ * @param tagConfig {Object} - the tag configuration
+ * @param payload {Object} - the payload object
+ * @returns - the mutated payload with the selected event properties added
+ */
+const addEventMappings = (evData, tagConfig, payload) => {
+  const returnObj = payload;
+  if (tagConfig.includeCommonEventProperties) {
+    COMMON_EVENT_PROPS.forEach((prop) => {
+      addProperty(prop, evData[prop], tagConfig, NEST.commonEvent, returnObj);
+    });
+  }
+  if (tagConfig.includeCommonUserProperties) {
+    const uKey = 'user_data';
+    if (evData[uKey]) {
+      addProperty(uKey, evData[uKey], tagConfig, NEST.userData, returnObj);
+    }
+  }
+  if (tagConfig.eventMappingRules) {
+    tagConfig.eventMappingRules.forEach((row) => {
+      const setVal = getEventData(row.key);
+      const mappedKey = row.mappedKey || row.key;
+      setFromPath(mappedKey, setVal, tagConfig, returnObj);
+    });
+  }
+  return returnObj;
+};
+
+/*
+ * Adds key-value pairs to the payload according to tag configuration.
+ *
+ * @param evData {Object} - the client event data
+ * @param tagConfig {Object} - the tag configuration
+ * @param payload {Object} - the payload object
+ * @returns - the mutated payload with the selected key-value pairs added
+ */
+const addRequestData = (tagConfig, payload) => {
+  const returnObj = payload;
+  if (tagConfig.additionalRequestData) {
+    tagConfig.additionalRequestData.forEach((row) => {
+      setFromPath(row.key, row.value, tagConfig, returnObj);
+    });
+  }
+  return returnObj;
+};
+
+/*
+ * Given the event data and the tag configuration, constructs the request body.
+ */
+const mkRequestPayload = (evData, tagConfig) => {
+  if (tagConfig.includeAll) {
+    return addRequestData(tagConfig, evData);
+  }
+
+  const target = {};
+  const requestPayload = cleanObject(
+    addRequestData(
+      tagConfig,
+      addEventMappings(
+        evData,
+        tagConfig,
+        parseSnowplowEvent(evData, tagConfig, target)
+      )
+    )
+  );
+
+  return tagConfig.inArray ? [requestPayload] : requestPayload;
+};
+
+/*
+ * Creates the HTTP request options according to the tag configuration.
+ */
+const mkRequestOptions = (tagConfig) => {
+  // requestTimeout input ensured by its validation rules
+  const timeout = makeInteger(tagConfig.requestTimeout);
+  const options = {
+    method: tagConfig.requestMethod.toUpperCase(),
+    timeout: timeout,
+    headers: {
+      'Content-Type': 'application/json',
```

Comment:
Perhaps worth documenting (in help text, or just in the docs) that this is already added to all requests?